programs/bpf_loader/src/lib.rs

```diff
@@ -248,11 +254,66 @@ pub fn create_executor_from_account(
         false, /* reject_deployment_of_broken_elfs */
     )?;
     if let Some(mut tx_executor_cache) = tx_executor_cache {
-        tx_executor_cache.set(*program.get_key(), executor.clone(), false);
+        tx_executor_cache.set(*program.get_key(), executor.clone(), false, false);
     }
     Ok((executor, Some(create_executor_metrics)))
 }
 
+macro_rules! deploy_program {
+    ($invoke_context:expr, $use_jit:expr, $program_id:expr,
+     $program_account:expr, $programdata_account:expr, $new_programdata:expr,
+     $first_drop:expr, $second_drop:expr $(,)?) => {{
+        let delay_visibility_of_program_deployment = $invoke_context
+            .feature_set
+            .is_active(&delay_visibility_of_program_deployment::id());
+
+        if delay_visibility_of_program_deployment {
+            // Make sure the old version is in the cache so that we don't
+            // load the new version from the database while the old one is still active
+            let mut get_or_create_executor_time = Measure::start("get_or_create_executor_time");
+            let result = create_executor_from_account(
```

Comment:
deploy might jit compile an program and place it in the cache even if it might never be called by this transaction?

Reply:
Yes, unfortunately that is necessary because we overwrite the original program in its account. So if it was called later there would be no way to load the old version, which is necessary so that the change stays invisible during the transaction.

Reply:
Or we fail any attempts to do so.  Aka, policy is you cannot call a program that you have previously upgraded in the same transaction

Reply:
Yes that would also be an option. However, this PR is only leading up to the replacement of the executor cache. The new one will extend the delay of visibility of changes way beyond the end of the transaction, to multiple slots.